src/protocol/ipa/mod.rs

```diff
@@ -0,0 +1,207 @@
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+enum Step {
+    ModulusConversionForMatchKeys,
+    GenSortPermutationFromMatchKeys,
+    ApplySortPermutation,
+    ComputeHelperBits,
+    AccumulateCredit,
+    PerformUserCapping,
+    AggregateCredit,
+}
+
+impl crate::protocol::Substep for Step {}
+
+impl AsRef<str> for Step {
+    fn as_ref(&self) -> &str {
+        match self {
+            Self::ModulusConversionForMatchKeys => "mod_conv_match_key",
+            Self::GenSortPermutationFromMatchKeys => "gen_sort_permutation_from_match_keys",
+            Self::ApplySortPermutation => "apply_sort_permutation",
+            Self::ComputeHelperBits => "compute_helper_bits",
+            Self::AccumulateCredit => "accumulate_credit",
+            Self::PerformUserCapping => "user_capping",
+            Self::AggregateCredit => "aggregate_credit",
+        }
+    }
+}
+
+/// # Errors
+/// Propagates errors from multiplications
+#[allow(dead_code)]
+pub async fn ipa<F>(
+    ctx: SemiHonestContext<'_, F>,
+    mk_shares: &[XorReplicated],
+    num_bits: u32,
+    other_inputs: Vec<Vec<Replicated<F>>>,
+    per_user_credit_cap: u32,
+) -> Result<Vec<AggregateCreditOutputRow<F>>, Error>
+where
+    F: Field,
+{
+    debug_assert_eq!(mk_shares.len(), other_inputs.len());
+
+    let local_lists = convert_all_bits_local(ctx.role(), mk_shares, num_bits);
+    let converted_shares = convert_all_bits(
+        &ctx.narrow(&Step::ModulusConversionForMatchKeys),
+        &local_lists,
+    )
+    .await
+    .unwrap();
+    let sort_permutation = generate_permutation_and_reveal_shuffled(
+        ctx.narrow(&Step::GenSortPermutationFromMatchKeys),
+        &converted_shares,
+        num_bits,
+    )
+    .await
+    .unwrap();
+
+    let combined_match_keys_and_sidecar_data = other_inputs
+        .into_iter()
+        .enumerate()
+        .map(|(row_index, mut sidecar_data)| {
+            for ith_bits in &converted_shares {
+                sidecar_data.push(ith_bits[row_index].clone());
+            }
+            sidecar_data
+        })
+        .collect::<Vec<_>>();
+
+    let sorted_rows = apply_sort_permutation(
+        ctx.narrow(&Step::ApplySortPermutation),
+        combined_match_keys_and_sidecar_data,
+        &sort_permutation,
+    )
+    .await
+    .unwrap();
+
+    let futures = zip(
+        repeat(ctx.narrow(&Step::ComputeHelperBits)),
+        sorted_rows.iter(),
+    )
+    .zip(sorted_rows.iter().skip(1))
+    .enumerate()
+    .map(|(i, ((ctx, row), next_row))| {
+        let record_id = RecordId::from(i);
+        async move { bitwise_equal(ctx, record_id, &row[3..], &next_row[3..]).await }
+    });
+    let helper_bits = try_join_all(futures).await?;
+
+    let attribution_input_rows = sorted_rows
+        .iter()
+        .enumerate()
+        .map(|(i, row)| {
+            let hb = if i == 0 {
+                Replicated::ZERO
+            } else {
+                helper_bits[i - 1].clone()
+            };
```

Comment:
You probably want to do `Some(Replicated::ZERO).into_iter().chain(helper_bits)` then zip that up with the rows.

Reply:
done!